src/inmanta/server/services/compilerservice.py

```diff
@@ -226,6 +227,84 @@ async def _run_compile_stage(self, name: str, cmd: list[str], cwd: str, env: dic
+        """
+        assert os.path.exists(self._project_dir)
+        project_dir = self._project_dir
+
+        # Eventual venv dir
+        venv_dir = os.path.join(project_dir, ".env")
+
+        # Versioned python dir
+        python_version = ".".join(platform.python_version_tuple()[0:2])
+        versioned_venv_dir = ".env-py" + python_version
+        versioned_venv_dir_full = os.path.join(project_dir, versioned_venv_dir)
+
+        # Determine the final action
+        # If we don't need to symlink, we return
+        do_create = False
+
+        if os.path.islink(venv_dir):
+            # We have a modern setup
+            try:
+                if os.path.samefile(venv_dir, versioned_venv_dir_full):
+                    # all good
+                    await self._info("Found existing venv")
+                    return None
+                else:
+                    # Wrong version
+                    os.unlink(venv_dir)
+                    await self._info(f"Creating new venv at {versioned_venv_dir_full}")
+                    do_create = True
+            except FileNotFoundError:
+                # Broken link or does not exist
+                await self._info(f"Creating new venv at {versioned_venv_dir_full}")
+                os.unlink(venv_dir)
+                do_create = True
+        elif not os.path.exists(venv_dir):
+            # Doesn't exist
+            do_create = True
+            await self._info(f"Creating new venv at {versioned_venv_dir_full}")
+        else:
+            # We have an older (< iso8) setup without symlink
+            virtual_env = VirtualEnv(venv_dir)
+            if virtual_env.exists():
+                try:
+                    virtual_env.can_activate()  # raises exception
+                    # version matches, move it to the correct folder
+                    os.rename(venv_dir, versioned_venv_dir_full)
+                    await self._info(f"Moving existing venv from {venv_dir} to {versioned_venv_dir_full}")
+                except VenvActivationFailedError:
+                    # Version doesn't match, move to alternative location
+                    os.rename(venv_dir, venv_dir + "_old")
+                    await self._info(f"Discarding existing venv from {venv_dir} to {venv_dir}_old, Creating new")
+                    do_create = True
+            else:
+                # otherwise broken
+                os.rename(venv_dir, venv_dir + "_old")
+                await self._info(f"Discarding existing venv from {venv_dir} to {venv_dir}_old, Creating new")
+                do_create = True
+
+        if do_create:
+            virtual_env = VirtualEnv(versioned_venv_dir_full)
+            virtual_env.init_env()
+
+        # Relative symlink is a bit exotic in python
+        try:
+            dir_fd = os.open(project_dir, os.O_RDONLY)
+            os.symlink(versioned_venv_dir, ".env", dir_fd=dir_fd)
+        finally:
+            if dir_fd:
+                os.close(dir_fd)
```

Comment:
I think relative symlinks are supported even without `dir_fd`:

```
sander@bedevere:~$ ll ~/downloads/dirtest/
total 52K
drwxr-xr-x   4 sander sander 4.0K Nov 20 15:59 ./
drwx------ 147 sander sander  36K Nov 20 15:08 ../
drwxr-xr-x   2 sander sander 4.0K Nov 20 15:09 test/
drwxr-xr-x   2 sander sander 4.0K Nov 20 15:08 test_old/
sander@bedevere:~$ python
Python 3.12.7 (main, Oct  1 2024, 11:15:50) [GCC 14.2.1 20240910] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.symlink("test", "/home/sander/downloads/dirtest/sl")
>>>
sander@bedevere:~$ ll ~/downloads/dirtest/
total 52K
drwxr-xr-x   4 sander sander 4.0K Nov 20 15:59 ./
drwx------ 147 sander sander  36K Nov 20 15:08 ../
lrwxrwxrwx   1 sander sander    4 Nov 20 15:59 sl -> test/
drwxr-xr-x   2 sander sander 4.0K Nov 20 15:09 test/
drwxr-xr-x   2 sander sander 4.0K Nov 20 15:08 test_old/
sander@bedevere:~$
```

Reply:
yes, but relative to what? (what if you are in a different folder?)